std/range/package.d

```diff
@@ -4756,21 +4755,28 @@ body
                                             cast(Value)(pastLast - (length - upper) * step),
                                             step);
         }
-        @property IndexType length() const
+        @property size_t length() const
         {
             if (step > 0)
             {
-                return unsigned((pastLast - current) / step);
+                return cast(size_t)unsigned((pastLast - current) / step);
             }
             else
             {
-                return unsigned((current - pastLast) / -step);
+                return cast(size_t)unsigned((current - pastLast) / -step);
             }
         }
 
         alias opDollar = length;
     }
 
+    if (step > 0)
+        assert(unsigned((end - begin) / step) <= size_t.max);
+    else if (step < 0)
+        assert(unsigned((end - begin) / -step) <= size_t.max);
+    //No else branch because if step = 0 it's either an error
+    //or begin == end and the range is empty
```

Comment:
Which comment? The one about maybe putting the `Result`'s constructor? `Result` doesn't even have an explicit constructor. But you're hitting the assertions under exactly the same circumstances whether they're right before calling `Result`'s constructor or in the precondition, and the precondition is a clearer place for preconditions to go. But regardless, separating the assertions out so that some are in the precondition and some are on the first lines of the function which are executed makes no sense, especially when you then have a comment to explain why you're not testing what was tested in the precondition. Just put them all together and make it clean.